### ISSUE TITLE
Upcoming v4 format

### ISSUE THREAD

**Issue by HeroicKatora** [pwsafe/pwsafe#300]
As noted [in this paper](https://www.cs.ox.ac.uk/files/6487/pwvault.pdf), not having to reencrypt the database on change of the master password (i.e. letting K, L remain unchanged) poses a potential security problem which cannot be easily dealt with by the user.

v3 of the database works around this problem by choosing different K, L on any save after a modification, including change of the master password. But v4 proposes a way for multiple, independent master passwords. In this format, one would be unable to ever change K, L without knowledge of all MPs, since one can only calculate KeyBlocks for known passwords.

Thus, the security of the database will differ substantially when adding/removing entries in multi password mode without asking all passwords at all time. What are your plans in order to deal with this perceived issue?

**Comment by HeroicKatora** [comment]
I have the following proposal, to alleviate the issue:

Algorithm
====
Each password Pi independently saves a unique encryption key Ki (i.e. has a block SALT | ITER | KW(Pi, Ki) and there exists only one section of wrapped K, L values. Its content will be explained below.

Then build a binary tree over all password entries where each node is used to execute some key-agreement algorithm. 
In each node save KW(Kil, Priv_l) | KW(Kir, Priv_r) | Pub_r | Pub_r | KW(Shared_lr, Knode) where 
 - Kil is the key of the left child (Ki if Pi is a password, or Knode if the child is itself a node)
 - Priv_l is the private data of the left child (e.g. the exponent of Diffie Hellman)
 - Pub_l is the public data of the left child (e.g. p^Priv_l for DH), analogue for the right node
 - Knode is an *unrelated random* bitstring, representing the key for this node
Finally, the internal key of the root node, Ktop, is to be used to decrypt the final values for K and L saved in their own section, as KW(Ktop, K) | KW(Ktop, L)

Considerations
====
Having n passwords, instead of needing O(n) space we need O(n log n) for the binary tree.
Decryption time will also be O(n log n) instead of O(n) (assuming constant ITER values for each key) since we need to crawl all log n nodes to the top.

Any key can update the data of any key-agreement scheme along its path to the root by storing a new pair of Priv_l, Pub_l and then reincrypting Knode or a replacement value for Knode under the new derived key. Since Ktop is also computable from any input key, we can also update K and L at all times, without knowledge about more than one key.

In order to expand the tree, we calculate the password chain from one or multiple given existing password(s) and then construct a new node, or a new root node, and update values accordingly. If we want to maintain a balanced tree to have log n depth, we either need a couple of passwords or, if we always append at the right, the last password added. When given all passwords, old and new, we can always build a balanced tree.

Security
====
I'll expand on this when I have time

**Comment by ronys** [comment]
Interesting - definitely worth study.

**Comment by HeroicKatora** [comment]
No problem, I'll update names for a few things above as I am now aware my subscripts are probably confusing as hell. 
The subscript _px shall be used to denote information directly related to the xth password.
Subscript _nx shall be used to denote information related to the node labeled x in the tree, for example PrivKey_n2 to denote a Private Key for node 2.

Just a short while to update the post above, meanwhile:
2. Shared_lr is a secret shared between the left and right child node.
3. The current scheme in v4 uses O(n) space, this will need O(n log n). This is worse but only by a log factor.
I'll give an example after the update to the indices

**Comment by ronys** [comment]
Thanks - I think I'm beginning to understand the gist of your proposal.

Give me a couple of weeks to get back to this.

**Comment by HeroicKatora** [comment]
Take your time. It would be quite useful if we had proof of any of algorithms _cryptographic_ properties, be it negative or positive proof. Ciphertext indistinguishability against other passwords (you can't decipher other stretched passwords) should probably be high on that list and I think we get it transitively through KW because we choose PrivData_pn at random [the former statement is not an attempted proof of any sort].

**Comment by dskrvk** [comment]
Wouldn't it make sense for this design to be written up in a paper and
reviewed by a wider expert community?

**Comment by HeroicKatora** [comment]
Yes it would, and I surely plan to. I just need to take some time writing it and find a place to distribute it.
I'll update this issue with a link when I publish something.

**Comment by HeroicKatora** [comment]
Reviving this issue, here's a more detailed writeup about this idea with better formatting.

https://www.hardmo.de/article/2018-01-14-shared-encrypted-knowledge.md

Note: I might make transparent edits, providing dates and clear markers
